### PR TITLE
EM_JS binaryen support

### PR DESCRIPTION
**Pull request by jgravelle-google**
Companion to https://github.com/kripken/emscripten/pull/6220

### CODE REVIEW

File: src/wasm-emscripten.cpp [hunk 1]
```diff
@@ -296,6 +297,65 @@ void EmscriptenGlueGenerator::generateJSCallThunks(
...
+    }
+  }
+  return code;
+}
+
+const char* stringAtAddr(Module& wasm,
+                         std::vector<Address> const& segmentOffsets,
+                         Address address) {
+  for (unsigned i = 0; i < wasm.memory.segments.size(); ++i) {
+    Memory::Segment &segment = wasm.memory.segments[i];
```
Comment (dschuff) [inline src/wasm-emscripten.cpp]: Apropos our recent formatting discussion, this should be `Memory::Segment& segment`

File: src/wasm-emscripten.cpp [hunk 2]
```diff
@@ -296,6 +297,65 @@ void EmscriptenGlueGenerator::generateJSCallThunks(
...
+                             std::vector<Address> const& segmentOffsets,
+                             Const* addrConst) {
+  auto address = addrConst->value.geti32();
+  const char* str = stringAtAddr(wasm, segmentOffsets, address);
+  if (!str) {
+    // If we can't find the segment corresponding with the address, then we
+    // omitted the segment and the address points to an empty string.
+    return escape("");
+  }
+  auto result = escape(str);
```
Comment (dschuff) [inline src/wasm-emscripten.cpp]: Why the temporary here?

Comment (dschuff) [reply src/wasm-emscripten.cpp]: Edit: oh I didn't realize that this was just moved from below. Still though :D

File: src/wasm-emscripten.cpp [hunk 3]
```diff
@@ -445,8 +453,50 @@ AsmConstWalker fixEmAsmConstsAndReturnWalker(Module& wasm) {
...
+      segmentOffsets(getSegmentOffsets(wasm)) { }
+
+  void visitFunction(Function* curr) {
+    if (!curr->name.startsWith(EM_JS_PREFIX.str)) {
+      return;
+    }
+    auto funcName = std::string(curr->name.stripPrefix(EM_JS_PREFIX.str));
+    Const* addrConst = curr->body->dynCast<Const>();
+    if (addrConst == nullptr) {
+      addrConst = curr->body->cast<Block>()->list[0]->cast<Const>();
```
Comment (dschuff) [inline src/wasm-emscripten.cpp]: This is parsing output from the compiler, right? I suppose we'll assert here if we unexpectedly get an emjs function that's not just a const? Maybe we should make a more explicit error check that also fires in release mode.

Comment (jgravelle-google) [reply src/wasm-emscripten.cpp]: Yeah, seems reasonable.

The idea here was that we should be the only ones generating these __em_js__* functions, so we should know if/when we ever change their format.
But if a user ever does prefix that, we might as well give a saner-sounding "hey no don't".

**Pull request merged**